File: tests/fixtures/c/control-flow.c
typedef void (*Callback)(void);

void control_flow(Callback (*factory)(void), Callback values[4]) {
    for (Callback loop = factory(); loop != 0; loop = factory()) {
        loop();
    }
    loop();

    for (int index = 0; index < 4; index++) {
        Callback item = values[index];
        item();
    }
    item();
}

File: tests/fixtures/rust/control-flow.rs
pub fn control_flow(
    source: impl Fn() -> Option<fn()>,
    values: Vec<fn()>,
    dynamic: Option<fn()>,
) {
    if let Some(selected) = source() {
        selected();
    } else {
        selected();
    }
    selected();

    while let Some(current) = source() {
        current();
    }
    current();

    for (_index, item) in values.into_iter().enumerate() {
        item();
    }
    item();

    match dynamic {
        Some(matched) if guard() => matched(),
        None => fallback(),
        _ => fallback(),
    }
    matched();

    let Some(finalized) = source() else {
        finalized();
        return;
    };
    finalized();

    let Point { callback, other: alias } = point();
    callback();
    alias();
}
